pa11ycrawler/pipelines.py

```diff
@@ -0,0 +1,108 @@
+# -*- coding: utf-8 -*-
+"""
+Item pipelines. Pipelines are enabled via the ITEM_PIPELINES setting.
+See: http://doc.scrapy.org/en/latest/topics/item-pipeline.html
+"""
+import os
+import json
+import subprocess
+import tempfile
+
+from scrapy.exceptions import DropItem
+
+from pa11ycrawler.settings_helpers import is_pa11y_setting
+
+
+class DuplicatesPipeline(object):
+    """
+    Ensures that we only process each URL once.
+    """
+    def __init__(self):
+        self.urls_seen = set()
+
+    def process_item(self, item, _spider):
+        """
+        Stops processing item if we've already seen this URL before.
+        """
+        if item['url'] in self.urls_seen:
+            raise DropItem("[Skipping] Duplicate url found: {}".format(item['url']))
+        else:
+            self.urls_seen.add(item['url'])
+            return item
+
+
+class Pa11yPipeline(object):
+    """
+    Runs the Pa11y CLI against `item['url']`, using the same headers used by
+    Scrapy. Stores the results in `item['results']` to be processed by
+    `spider.a11y_reporter`.
+    """
+    def __init__(self):
+        self.pa11y_settings = None
+
+    def get_pa11y_settings(self, spider):
+        """
+        Sets `self.pa11y_settings` to the list of flags that should
+        be passed to the Pa11y CLI.
+        """
+        settings = []
+        for key, value in spider.settings.attributes.iteritems():
+            if is_pa11y_setting(key) and value.value:
+                flag_name = key.replace('_', '-').split('PA11Y-')[1].lower()
+                flag = '--{}="{}"'.format(flag_name, value.value)
+                settings.append(flag)
+
+        self.pa11y_settings = settings
+        return settings
+
+    def _gen_pa11y_config(self, headers):
+        """
+        Creates tempfile with pa11y configuration.
+        This includes any headers that need to be kept.
+        Returns the file path of the new file.
+        """
+        config = json.dumps({
+            "page": {
+                "headers": headers,
+            },
+        })
+
+        pa11y_config_fp = tempfile.mkstemp(suffix='.json')[1]
```

Comment:
This will put it into some random directory, right? Should it go into a known location and be deleted instead?

No strong preference but wanted to ask.